src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp

```diff
@@ -820,31 +821,91 @@ HRESULT CorProfiler::ProcessReplacementCalls(
 
-      // insert the opcode and signature token as
-      // additional arguments for the wrapper method
-      //
-      // IMPORTANT: Conditional branches may jump to this call instruction, so
-      // we cannot add the argument-loading instructions BEFORE this instruction,
-      // otherwise this will surface in an InvalidProgramException.
-      // Either begin loading the additional arguments starting with this ILInstr
-      // structure or make the current ILInstr a no-op and do all of the argument
-      // loading after this instruction.
+      // Begin IL Modification
       ILRewriterWrapper rewriter_wrapper(&rewriter);
       rewriter_wrapper.SetILPosition(pInstr);
+
+      // IL Modification #1: Replace original method call with a NOP, so that all original
+      //                     jump targets resolve correctly and we correctly populate the
+      //                     stack with additional arguments
+      //
+      // IMPORTANT: Conditional branches may jump to the original call instruction which
+      // resulted in the InvalidProgramException seen in
+      // https://github.com/DataDog/dd-trace-dotnet/pull/542. To avoid this, we'll do
+      // the rest of our IL modifications AFTER this instruction.
       auto original_methodcall_opcode = pInstr->m_opcode;
       pInstr->m_opcode = CEE_NOP;
+      pInstr = pInstr->m_pNext;
+      rewriter_wrapper.SetILPosition(pInstr);
 
-      // replace with a non-virtual call (CALL) to the instrumentation wrapper
-      // always use CALL because the wrappers methods are all static
-      rewriter_wrapper.CallMemberAfter(wrapper_method_ref, false);
-      rewriter_wrapper.SetILPosition(pInstr->m_pNext);
+      // IL Modification #2: Conditionally box System.Threading.CancellationToken
+      //                     if it is the last argument in the target method.
+      //
+      // If the last argument in the method signature is of the type
+      // System.Threading.CancellationToken (a struct) then box it before calling our
+      // integration method. This resolves https://github.com/DataDog/dd-trace-dotnet/issues/662
+      // which reproduces when the CLR automatically boxes System.Threading.CancellationToken
+      // when the original target method was in System.Data and we're using the 32-bit profiler
+      // on .NET Framework.
+      //
+      // Currently, all integrations that use System.Threading.CancellationToken (a struct)
+      // have the argument as the last argument in the signature (lucky us!).
+      // For now, we'll do the following:
+      //   1) Get the method signature of the original target method
+      //   2) Read the signature until the final argument type
+      //   3) If the type begins with `ELEMENT_TYPE_VALUETYPE`, uncompress the compressed type token that follows
+      //   4) If the type token represents System.Threading.CancellationToken, emit a 'box <type_token>' IL instruction before calling our wrapper method
```

Comment:
Sorry I didn't suggest this earlier, but it just occurred to me: should we consider replacing the `CancellationToken` on the stack with `CancellationToken.m_source`? That would avoid having to box/unbox on every async call we instrument.

Since the `CancellationToken` is already on the stack, I think it should be easy enough to emit a [`ldfld`](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.emit.opcodes.ldfld) opcode instead of a `box`.

Reply:
Maybe merge this fix and we can do that as performance improvement later.

Reply:
That very well could work, but I think the procedure of boxing the variable is the more correct thing to do and it is more generalizable if we have to extend this to other value types later on.